openupgrade_scripts/scripts/sale_loyalty/16.0.1.0/pre-migration.py

```diff
@@ -0,0 +1,299 @@
+            loyalty_card_id INT
+        )
+        """,
+    )
+    openupgrade.logged_query(
+        env.cr,
+        """
+        INSERT INTO loyalty_card_sale_order_rel (sale_order_id, loyalty_card_id)
+        SELECT sales_order_id, id
+        FROM loyalty_card
+        """,
+    )
+
+
+def fill_code_enabled_rule_ids_from_sale_order(env):
+    """This is an m2m field, the relationship table is created in case it does not
+    exist and filled with the data. This field relates the id of a rule to the id of
+    a sales order to which it has been applied, provided that the rule belongs to an
+    applicable promotion with code."""
+    openupgrade.logged_query(
+        env.cr,
+        """
+        CREATE TABLE IF NOT EXISTS loyalty_rule_sale_order_rel (
+            sale_order_id INT,
+            loyalty_rule_id INT
+        )
+        """,
+    )
+    openupgrade.logged_query(
+        env.cr,
+        """
+        INSERT INTO loyalty_rule_sale_order_rel (sale_order_id, loyalty_rule_id)
+        SELECT lc.order_id, lr.id
+        FROM loyalty_card lc
+        JOIN loyalty_rule lr ON lc.program_id = lr.program_id
+        JOIN loyalty_program lp ON lr.program_id = lp.id
+        WHERE (lr.mode = 'with_code' OR lp.program_type = 'gift_card')
+        """,
+    )
+
+
+def generate_sale_order_coupon_points(env):
+    openupgrade.logged_query(
+        env.cr,
+        """
+        CREATE TABLE IF NOT EXISTS sale_order_coupon_points (
+            coupon_id INT,
+            order_id INT,
+            points FLOAT,
+            create_uid INT,
+            write_uid INT,
+            create_date DATE,
+            write_date DATE
+        )
+        """,
+    )
+    openupgrade.logged_query(
+        env.cr,
+        """
+        INSERT INTO sale_order_coupon_points (
+            coupon_id, order_id, points, create_uid, write_uid, create_date, write_date
+        )
+        SELECT
+            id AS coupon_id,
+            order_id,
+            1 AS points,
+            create_uid,
+            write_uid,
+            create_date,
+            write_date
+        FROM loyalty_card;
+        """,
+    )
+
+
+def update_loyalty_program_data(env):
+    # Set sale_ok default values
+    openupgrade.logged_query(
+        env.cr,
+        """
+        ALTER TABLE loyalty_program
+        ADD COLUMN IF NOT EXISTS sale_ok BOOLEAN
+        """,
+    )
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE loyalty_program
+        SET sale_ok = true
+        """,
+    )
+
+
+def update_sale_order_line_data(env):
+    # Fill reward_id field
+    openupgrade.logged_query(
+        env.cr,
+        """
+        ALTER TABLE sale_order_line
+        ADD COLUMN IF NOT EXISTS reward_id INT
+        """,
+    )
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE sale_order_line sol
+        SET reward_id = lr.id
+        FROM loyalty_reward lr
+        WHERE sol.product_id = lr.discount_line_product_id
```

Comment:
This seems a weak way to link rewards only through product. Isn't there another option?

Reply:
discount_line_product_id is a unique product.product that is created for that reward_id, so it is the only link to get that reward_id. In addition, each program has its own rewards, so there can be no confusion as they are unique data. 
![image](https://github.com/OCA/OpenUpgrade/assets/118818446/ee00a5ce-2414-4086-ad26-6c1128f0903b)